pkg/misconf/scanner.go

```diff
@@ -214,11 +301,38 @@ func (s *Scanner) Scan(ctx context.Context, files []types.File) ([]types.Misconf
 				}
 			}
 		}
-	}
 
-	var misconfs []types.Misconfiguration
-	for t, scanner := range s.scanners {
-		results, err := scanner.ScanFS(ctx, mapMemoryFS[t], ".")
+		var results scan.Results
+		var rootDir string
+		var err error
+
+		switch t {
+		case types.Terraform:
+			rootDir, err = getRootDir(findCommonPrefix(relevantFiles))
+			if err != nil {
+				log.Logger.Debugf("failed to find config root path: %s", err)
+				results, err = scanner.ScanFS(ctx, mapMemoryFS[t], ".")
+				if err != nil {
+					return nil, xerrors.Errorf("scanfs for %s scan from memoryfs failed: %w", t, err)
+				}
+			} else {
+				// Support Windows paths
+				if volume := filepath.VolumeName(rootDir); volume != "" {
+					rootDir = strings.TrimPrefix(filepath.ToSlash(rootDir), volume+"/")
+				}
+
+				results, err = scanner.ScanFS(ctx, extrafs.OSDir("/"), rootDir)
```

Comment:
> `extrafs.OSDir` says
> 
> > Go does not currently support symlinks in io/fs
> 
> https://github.com/aquasecurity/defsec/blob/a071f4812c575431b38ce741c2138c40cfcc423/pkg/extrafs/extrafs.go#L9-L12
> 
> https://github.com/aquasecurity/defsec/blob/a071f4812c575431b38ce741c2138c40cfccc423/pkg/extrafs/extrafs.go#L9-L12
> 
> Is it correct? DirFS seems to follow symlinks. Am I missing something? https://pkg.go.dev/os#DirFS

Hmm... I don't know the origin of that comment but I do see os.DirFS does follow symlinks as you said. Maybe that comment/commit was added pre go 1.16 and isn't relevant anymore? 🤷🏼 


> And it breaks container/VM image scanning since `os.DirFS` refers to the local filesystem. Putting Terraform files into container/VM images is rare, but we have actually seen it in the past. That is why we use `memoryfs` instead of the OS filesystem.
> 

Oh interesting. We should have a test for this. Do we have any integration tests that test out container/VM images? Maybe we can add a symlinked terraform file in there to have this covered. 

> IIUC, the problem here is `memoryfs` and `os.DirFS` don't allow accessing outside the root directory. I have a plan to replace `memoryfs` with `mapfs`. What if allowing `mapfs` to access the outside file?

Yes that's the problem. I assume you are talking [this](https://github.com/aquasecurity/trivy/blob/main/pkg/mapfs/fs.go) `mapfs` right? Does it support symlinks today? I only see that it implements these https://github.com/aquasecurity/trivy/blob/main/pkg/mapfs/fs.go#L18-L22 - I assumed you would need to implement something like this: https://github.com/aquasecurity/defsec/blob/a071f4812c575431b38ce741c2138c40cfccc423/pkg/extrafs/extrafs.go#L48-L54?


> 
> I'll raise a PR shortly. We can discuss the solution after that.

Sure, add me to the review so I'll take a look at it as well.

Reply:
> Does it support symlinks today?

It uses `os.Open()` under the hood, which follows symlinks. What else do we need?
https://github.com/aquasecurity/trivy/blob/cc18f92cf39534a2220883722fb26dc1a9e27fda/pkg/mapfs/file.go#L68-L69

Reply:
In theory, tthat should be all. The key thing would be to be able to support paths outside the `memfs`, including relative paths, which is what this PR tries to do with `extrafs`.